File: bootstrap/src/main/java/net/dontdrinkandroot/wicket/bootstrap/component/item/RepeatingDropdownItem.java
/*
 * Copyright (C) 2012-2017 Philip Washington Sorst <philip@sorst.net>
 * and individual contributors as indicated
 * by the @authors tag.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.dontdrinkandroot.wicket.bootstrap.component.item;

import net.dontdrinkandroot.wicket.bootstrap.component.dropdown.DropdownMenu;
import org.apache.wicket.markup.repeater.RepeatingView;
import org.apache.wicket.model.IModel;

/**
 * {@link DropdownItem} that is populated via a {@link RepeatingView}.
 *
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public abstract class RepeatingDropdownItem<T> extends DropdownItem<T>
{
    public RepeatingDropdownItem(String id, IModel<String> labelModel)
    {
        this(id, labelModel, null);
    }

    public RepeatingDropdownItem(String id, IModel<String> labelModel, IModel<T> model)
    {
        super(id, labelModel, model);
    }

    @Override
    protected DropdownMenu createDropdownMenu(String id)
    {
        return new DropdownMenu(id)
        {
            @Override
            protected void populateItems(RepeatingView itemView)
            {
                RepeatingDropdownItem.this.populateItems(itemView);
            }
        };
    }

    protected abstract void populateItems(RepeatingView itemView);
}

File: bootstrap/src/main/kotlin/net/dontdrinkandroot/wicket/bootstrap/component/item/RepeatingDropdownItem.kt
package net.dontdrinkandroot.wicket.bootstrap.component.item

import net.dontdrinkandroot.wicket.bootstrap.component.dropdown.DropdownMenu
import net.dontdrinkandroot.wicket.bootstrap.css.DropdownAlignment
import org.apache.wicket.behavior.Behavior
import org.apache.wicket.markup.repeater.RepeatingView
import org.apache.wicket.model.IModel
import org.apache.wicket.model.Model

/**
 * [DropdownItem] that is populated via a [RepeatingView].
 */
abstract class RepeatingDropdownItem<T>(
    id: String,
    model: IModel<T>? = null,
    labelModel: IModel<String>,
    private val dropdownAlignment: IModel<DropdownAlignment?> = Model(null),
    vararg linkBehaviors: Behavior,
) : DropdownItem<T>(id, model, labelModel, *linkBehaviors) {

    override fun createDropdownMenu(id: String) = object : DropdownMenu(id, dropdownAlignment) {
        override fun populateItems(itemView: ItemView) {
            this@RepeatingDropdownItem.populateItems(itemView)
        }
    }

    abstract fun populateItems(itemView: ItemView)
}

fun <T> repeatingDropdownItem(
    id: String,
    model: IModel<T>? = null,
    labelModel: IModel<String>,
    dropdownAlignment: IModel<DropdownAlignment?> = Model(null),
    vararg behaviors: Behavior,
    populateItemsHandler: ItemView.() -> Any?
) = object : RepeatingDropdownItem<T>(id, model, labelModel, dropdownAlignment, *behaviors) {
    override fun populateItems(itemView: ItemView) {
        populateItemsHandler(itemView)
    }
}

fun repeatingDropdownItem(
    id: String,
    labelModel: IModel<String>,
    dropdownAlignment: IModel<DropdownAlignment?> = Model(null),
    vararg behaviors: Behavior,
    populateItemsHandler: ItemView.() -> Any?
) = object : RepeatingDropdownItem<Void>(id, null, labelModel, dropdownAlignment, *behaviors) {
    override fun populateItems(itemView: ItemView) {
        populateItemsHandler(itemView)
    }
}